symfit/core/fit.py

```diff
@@ -1244,13 +1271,40 @@ def __init__(self, model, *ordered_data, **named_data):
         objective = named_data.pop('objective')
         minimizer = named_data.pop('minimizer')
         constraints = named_data.pop('constraints')
+        # Should be a list of Constraint objects
         constraints = [] if constraints is None else constraints
-        # List of Constraint objects
-        if not isinstance(model, BaseModel):
-            model = Model(model)
+
+        # Initiate self.model as an instance of BaseModel if it isn't already
+        if isinstance(model, BaseModel):
+            self.model = model
+        else:
+            self.model = Model(model)
+
+        # Select objective function to use. Has to be done before calling
+        # super.__init__
+        self.objective = self._determine_objective(self.model, objective,
+                                                   minimizer)
         self.constraints = self._init_constraints(constraints=constraints,
-                                                  model=model)
-        super(Fit, self).__init__(model, *ordered_data, **named_data)
+                                                  model=self.model)
+        try:
+            super(Fit, self).__init__(model, *ordered_data, **named_data)
+        except TypeError as err:
+            # There is one possible exception: if the objective was determined
+            # to be a MinimizeModel, but the user actually provided data for the
+            # dependent vars. In that case, they actually want a LeastSquare so
+            # we retry.
```

Comment:
Doesn't this mean that objective auto-detection should be smarter, rather than trying to fix it up afterwards?

Reply:
Yeah I agree in principle, but I have been unable to do that so far. The main problem is that the determination depends on both the objective and the data. `TakesData` does not know about objectives, it only matches the data to the model. I feel like it isn't right to pass the objective to the `TakesData` super, but if we don't then the signature will not be satisfied and thus raise an error.

One way around this might be to have a seperate class/function which checks the signature based on the model, data and optionally the objective, which both TakesData and FIt can call in their own way.

Reply:
I'm slightly confused on who calls what now, but that's ok.
My initial goal would be to keep all the "intelligence" in Fit, which knows data and model, and can therefor select the appropriate Objective. Who is missing information then?

Reply:
This is solved now.